### ISSUE TITLE
False positive arguments-differ when overwritten args kwargs are expanded

### ISSUE THREAD

**Issue by peldszus** [PyCQA/pylint#3812]
The `arguments-differ` check complains, when args/kwargs of the base method are expanded in overwritten methods.

I would assume that this should be considered OK, as all overwritten signatures are subsumed by the base methods signature.

This seems to be related but not exactly similar to #3672, #1553, and #1556.

### Steps to reproduce
```python
class Base:

    def run(self, *args, **kwargs):
        pass


class ExpandArguments(Base):

    def run(self, argument, *args, **kwargs):
        pass


class ExpandKeywords(Base):

    def run(self, *args, keyword=None, **kwargs):
        pass


class ExpandArgumentsAndKeywords(Base):

    def run(self, argument, *args, keyword=None, **kwargs):
        pass
```

### Current behavior
`arguments-differ` is emitted:
```
...
test.py:9:4: W0221: Parameters differ from overridden 'run' method (arguments-differ)
...
test.py:15:4: W0221: Parameters differ from overridden 'run' method (arguments-differ)
...
test.py:21:4: W0221: Parameters differ from overridden 'run' method (arguments-differ)
```

### Expected behavior
`arguments-differ` is not emitted.

### pylint --version output
```
pylint 2.6.0
astroid 2.4.2
Python 3.6.9 (default, Jul 17 2020, 12:50:27) 
[GCC 8.4.0]
```

**Comment by AWhetter** [comment]
`ExpandArguments` genuinely has a different signature because it's valid to pass no arguments to `Base.run()` but at least one argument is required to `ExpandArguments.run()`.

For `ExpandKeywords`, it's difficult to extrapolate what the developer intended when writing `**kwargs` in `Base.run()`.
Maybe it's a signature similar to that of `dict()` where all keyword arguments are stored somewhere. Adding `keyword=None` to the signature means that it's now impossible to store "keyword" as a key in the dictionary for example.
Let's say `ExpandKeywords.run()` is calling through to `super().run(*args, keyword=keyword, **kwargs)` and  `Base.run()` is doing something depending on whether `keyword` is passed as an argument or not. That can't be figured out anymore because `keyword` is always set to None. Maybe the author of `ExpandKeywords` intended that, but maybe not.
Overall, it's difficult for pylint to make the right call on whether this is an issue or not. I think this will always be a case where a developer has to get involved to verify what's going on, and so pylint raising the warning is the right thing to alert the developer to double check what they're doing? If it is the right thing, this seems like a valid place for a `# pylint: disable=arguments-differ` pragma.

I'm going to close this issue for now but let me know if you have any further thoughts on this.

**Comment by peldszus** [comment]
Hi Ashley!

Thanks for taking the time to look into this and apologies for taking so long to respond.

I agree with your points. It turned out, I interpreted this rule too much unidirectional (only considering compatiblility of the overwritten method with the base method, not the other way around).

FYI a bit background: The case why I came up with this (HTTP verb methods in Tornado request handlers) was that the base class trivially implements the method and is able to catch all args and kwargs. You could then overwrite this, to handle any args and kwargs more specifically. So everything that you could pass to the overwritten method could also be passed to the base method (where it would be handled trivially). But you are right that this is not true vice versa.

Raising this warning is thus right and a good trigger to force me to check, and if legit to put a prama. :)

**Comment by embedl-axel** [comment]
I am wondering about a slightly different combinations of signatures:

```
class NoArgsBase:
    def run(self, **kwargs):
        pass

class ExpandedNoArgs(Base):
    def run(self,  keyword=None, **kwargs):
        pass

class KeywordOnlyBase:
    def run(self, arg0, arg1, *, **kwargs):
        pass

class ExpandKeywordOnlyBase(Base):
    def run(self, arg0, arg1, *, keyword=None, **kwargs):
        pass
```

Is there a reason why pylint would not be able to determine that the arguments of the child class are compatible with those of the parent class?

**Comment by cdce8p** [comment]
> I am wondering about a slightly different combinations of signatures:
> ...
> Is there a reason why pylint would not be able to determine that the arguments of the child class are compatible with those of the parent class?

Can you post your pylint output as well as `pylint --version`?

Note that your code contains a SyntaxError:
```diff
 class KeywordOnlyBase:
-    def run(self, arg0, arg1, *, **kwargs):
+    def run(self, arg0, arg1, **kwargs):
         pass
```
